src/dbnode/persist/fs/msgpack/decoder.go

```diff
@@ -76,16 +80,22 @@ func newDecoder(legacy legacyEncodingOptions, opts DecodingOptions) *Decoder {
 		reader:            reader,
 		dec:               msgpack.NewDecoder(reader),
 		legacy:            legacy,
+		streamWithDigest:  newDecoderStreamWithDigest(nil),
 	}
 }
 
 // Reset resets the data stream to decode from
 func (dec *Decoder) Reset(stream DecoderStream) {
 	dec.reader = stream
 
+	unwrappedStream := stream
+	if streamWithDigest, ok := stream.(DecoderStreamWithDigest); ok {
+		unwrappedStream = streamWithDigest.wrappedStream()
```

Comment:
Hm, interesting. What's this required for? Wouldn't we always do the wrapping in the decoder, not defer to the stream passed in to have that feature enabled?

Reply:
It's true that we always do the wrapping in the decoder. However, to enable the wrapped stream for use, we call `decoder.Reset(wrappedStream)` in `wrapWithStreamWithDigest`. This logic here is done so that any `byteReader` that's currently set doesn't get niled out when we call `decoder.Reset(wrappedStream)` as the wrapped stream implements the `DecoderStream` interface, but not the `ByteStream` interface. This ensures that the `byteReader` is still available for use for those special cases where we want to access the backing bytes directly (fyi, this is also the case where `DecoderStreamWithDigest#Capture` is used: https://github.com/m3db/m3/pull/2468/files#diff-a4220f1ade552cd435cedf33723b0f81R686)

We could avoid this logic in `Reset` by not calling `decoder.Reset` from `wrapWithStreamWithDigest` and just setting `decoder.reader` and calling `decoder.dec.Reset` directly. Thinking about it more, maybe that's less convoluted?